datafusion/core/src/datasource/file_format/parquet.rs

```diff
@@ -543,6 +574,172 @@ async fn fetch_statistics(
+            )),
+            FileWriterMode::PutMultipart => {
+                let (_, multipart_writer) = object_store
+                    .put_multipart(&object.location)
+                    .await
+                    .map_err(DataFusionError::ObjectStore)?;
+                let writer = AsyncArrowWriter::try_new(
+                    multipart_writer,
+                    self.config.output_schema.clone(),
+                    10485760,
+                    Some(parquet_props),
+                )?;
+                Ok(writer)
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl DataSink for ParquetSink {
+    async fn write_all(
+        &self,
+        mut data: Vec<SendableRecordBatchStream>,
+        context: &Arc<TaskContext>,
+    ) -> Result<u64> {
+        let num_partitions = data.len();
+        let parquet_props = self.parquet_writer_props_from_context(context);
+
+        let object_store = context
+            .runtime_env()
+            .object_store(&self.config.object_store_url)?;
+
+        // Construct writer for each file group
+        let mut writers = vec![];
+        match self.config.writer_mode {
+            FileWriterMode::Append => {
+                return plan_err!(
+                    "Parquet format does not support appending to existing file!"
+                )
+            }
+            FileWriterMode::Put => {
+                return Err(DataFusionError::NotImplemented(
+                    "Put Mode is not implemented for ParquetSink yet".into(),
+                ))
+            }
+            FileWriterMode::PutMultipart => {
+                // Currently assuming only 1 partition path (i.e. not hive-style partitioning on a column)
+                let base_path = &self.config.table_paths[0];
+                // Uniquely identify this batch of files with a random string, to prevent collisions overwriting files
+                let write_id = Alphanumeric.sample_string(&mut rand::thread_rng(), 16);
+                for part_idx in 0..num_partitions {
+                    let file_path = base_path
+                        .prefix()
+                        .child(format!("/{}_{}.parquet", write_id, part_idx));
+                    let object_meta = ObjectMeta {
+                        location: file_path,
+                        last_modified: chrono::offset::Utc::now(),
+                        size: 0,
+                        e_tag: None,
+                    };
+                    let writer = self
+                        .create_writer(
+                            object_meta.into(),
+                            object_store.clone(),
+                            parquet_props.clone(),
+                        )
+                        .await?;
+                    writers.push(writer);
+                }
+            }
+        }
+
+        let mut row_count = 0;
+        // TODO parallelize serialization accross partitions and batches within partitions
```

Comment:
I am digging the fact all the writers look quite similar -- we'll be able to sort out parallelizing the writes in one place and use it for all the formats ❤️